lib/code-caller-python.js

```diff
@@ -124,27 +123,42 @@ class PythonCaller {
         debug(`exit constructor(), state: ${String(this.state)}, uuid: ${this.uuid}`);
     }
 
-    call(file, fcn, args, options, callback) {
+    prepareForCourse(course, callback) {
+        debug(`enter prepareForCourse(), state: ${String(this.state)}, uuid: ${this.uuid}`);
+        this.course = course;
+        debug(`exit prepareForCourse(), state: ${String(this.state)}, uuid: ${this.uuid}`);
+        callback(null);
+    }
+
+    call(type, directory, file, fcn, args, callback) {
         debug(`enter call(), state: ${String(this.state)}, uuid: ${this.uuid}`);
         if (!this._checkState([CREATED, WAITING])) return callback(new Error('invalid PythonCaller state'));
 
         if (this.state == CREATED) {
             this._startChild();
         }
 
-        const localOptions = _.defaults(options, {
-            cwd: __dirname,
-            paths: [],
-            timeout: 20000, // FIXME: this number (equivalent to 20 seconds) should not have to be this high
-        });
-        const callData = {
-            file, fcn, args,
-            cwd: localOptions.cwd,
-            paths: localOptions.paths,
-        };
+        let cwd;
+        const paths = [path.join(__dirname, '..', 'python')];
+        if (type === 'question') {
+            cwd = path.join(this.course.path, 'questions', directory);
+            paths.push(path.join(this.course.path, 'serverFilesCourse'));
+        } else if (type === 'course-element') {
+            cwd = path.join(this.course.path, 'elements', directory);
+            paths.push(path.join(this.course.path, 'serverFilesCourse'));
+        } else if (type === 'core-element') {
+            cwd = path.join(__dirname, '..', 'elements', directory);
+        } else if (type === 'restart') {
+            // Doesn't need a working directory
+        } else {
+            callback(new Error(`Unknown function call type: ${type}`));
+            return;
+        }
+
+        const callData = { file, fcn, args, cwd, paths };
         const callDataString = JSON.stringify(callData);
         this.callback = callback;
-        this.timeoutID = setTimeout(this._timeout.bind(this), localOptions.timeout);
+        this.timeoutID = setTimeout(this._timeout.bind(this), 5000);
```

Comment:
Maybe add a config variable for this? I guess we used to allow this to be set per-question (or per-element). My thinking was that we should generally enforce a short timeout on question code, but it might be good to have an escape hatch to allow some question code to run longer.

Reply:
This now picks up `config.questionTimeoutMilliseconds`.